java-source/src/test/java/com/example/contract/IOUFlowTests.java

```diff
@@ -0,0 +1,219 @@
+//        IOUState state = new IOUState(
+//                new IOU(1),
+//                a.info.getLegalIdentity(),
+//                c.info.getLegalIdentity(),
+//                new IOUContract());
+//        ExampleFlow.Initiator flow = new ExampleFlow.Initiator(state, b.info.getLegalIdentity());
+//        ListenableFuture<SignedTransaction> future = a.getServices().startFlow(flow).getResultFuture();
+//        net.runNetwork(-1);
+//
+//        assertFails(future.get);
+//    }
+
+    @Test
+    public void flowRecordsATransactionInBothPartiesVaults() throws InterruptedException, ExecutionException {
+        IOUState state = new IOUState(
+                new IOU(1),
+                a.info.getLegalIdentity(),
+                b.info.getLegalIdentity(),
+                new IOUContract());
+        ExampleFlow.Initiator flow = new ExampleFlow.Initiator(state, b.info.getLegalIdentity());
+        ListenableFuture<SignedTransaction> future = a.getServices().startFlow(flow).getResultFuture();
+        net.runNetwork(-1);
+        SignedTransaction signedTx = future.get();
+
+        databaseTransaction(a.database, it -> {
+            assertEquals(signedTx, a.storage.getValidatedTransactions().getTransaction(signedTx.getId()));
+            return null;
+        });
+
+        databaseTransaction(b.database, it -> {
+            assertEquals(signedTx, b.storage.getValidatedTransactions().getTransaction(signedTx.getId()));
+            return null;
+        });
+    }
+
+    @Test
+    public void recordedTransactionHasNoInputsAndASingleOutputTheInputIOU() throws InterruptedException, ExecutionException {
+        IOUState inputState = new IOUState(
+                new IOU(1),
+                a.info.getLegalIdentity(),
+                b.info.getLegalIdentity(),
+                new IOUContract());
+        ExampleFlow.Initiator flow = new ExampleFlow.Initiator(inputState, b.info.getLegalIdentity());
+        ListenableFuture<SignedTransaction> future = a.getServices().startFlow(flow).getResultFuture();
+        net.runNetwork(-1);
+        SignedTransaction signedTx = future.get();
+
+        databaseTransaction(a.database, it -> {
+            SignedTransaction recordedTx = a.storage.getValidatedTransactions().getTransaction(signedTx.getId());
+            List<TransactionState<ContractState>> txOutputs = recordedTx.getTx().getOutputs();
+            assert(txOutputs.size() == 1);
+
+            IOUState recordedState = (IOUState) txOutputs.get(0).getData();
+            assertEquals(recordedState.getIOU().getValue(), inputState.getIOU().getValue());
+            assertEquals(recordedState.getSender(), inputState.getSender());
+            assertEquals(recordedState.getRecipient(), inputState.getRecipient());
+            assertEquals(recordedState.getLinearId(), inputState.getLinearId());
+            return null;
```

Comment:
That's fine, it's just this ugly null you have to return because this method is better suited to Kotlin. I'm just thinking that if we were in pure Java then the standard thing to do would be what I described above. 

We could of course add another method which takes in a `Consumer` and this return null thing would go away.